src/Microsoft.AspNetCore.OData/Query/Expressions/BinderExtensions.cs

```diff
@@ -355,5 +361,93 @@ public static IQueryable ApplyBind(this ISearchBinder binder, IQueryable source,
             Expression searchExp = binder.BindSearch(searchClause, context);
             return ExpressionHelpers.Where(source, searchExp, context.ElementClrType);
         }
+
+        /// <summary>
+        /// Translate an OData $apply parse tree represented by <see cref="TransformationNode"/> to
+        /// an <see cref="Expression"/> and applies it to an <see cref="IQueryable"/>.
+        /// </summary>
+        /// <param name="binder">An instance of <see cref="IAggregationBinder"/>.</param>
+        /// <param name="source">The original <see cref="IQueryable"/>.</param>
+        /// <param name="transformationNode">The OData $apply parse tree.</param>
+        /// <param name="context">An instance of the <see cref="QueryBinderContext"/> containing the current query context.</param>
+        /// <param name="resultClrType">The type of wrapper used to create an expression from the $apply parse tree.</param>
+        /// <returns>The applied result.</returns>
+        public static IQueryable ApplyBind(this IAggregationBinder binder, IQueryable source, TransformationNode transformationNode, QueryBinderContext context, out Type resultClrType)
+        {
+            if (binder == null)
+            {
+                throw Error.ArgumentNull(nameof(binder));
+            }
+
+            if (source == null)
+            {
+                throw Error.ArgumentNull(nameof(source));
+            }
+
+            if (transformationNode == null)
+            {
+                throw Error.ArgumentNull(nameof(transformationNode));
+            }
+
+            if (context == null)
+            {
+                throw Error.ArgumentNull(nameof(context));
+            }
+
+            // Ensure that the flattened properties are populated for the current query context.
+            context.EnsureFlattenedProperties(context.CurrentParameter, source);
+
+            // ApplyBind may be called multiple times if there are multiple groupby transformations
+            // e.g., $apply=groupby((a,b),aggregate(c))/groupby((d),aggregate(e))
+            // In this case, the first groupby will be applied to the original source,
+            // and the second groupby will be applied to the result of the first groupby
+            // There would be no reason to flatten the properties again if they were already flattened
+            if (!(context.FlattenedProperties?.Any() == true))
```

Comment:
Why not just saying:
```suggestion
            if (context.FlattenedProperties.Any() != true)
```